include/AST/Decl/FunctionDecl.hpp

```diff
@@ -23,33 +23,34 @@ class FunctionDecl final
     using TrailingParams = llvm::TrailingObjects<FunctionDecl, ParamDecl>;
     llvm::StringRef _name;
     glu::types::FunctionTy *_type;
-    CompoundStmt _body;
+    CompoundStmt *_body;
 
     unsigned _numParams;
 
     // Method required by llvm::TrailingObjects to determine the number
     // of trailing objects.
-    size_t numTrailingObjects(typename TrailingParams::OverloadToken<ParamDecl>)
-        const
+    size_t
+        numTrailingObjects(typename TrailingParams::OverloadToken<ParamDecl>) const
     {
         return _numParams;
     }
 
     FunctionDecl(
-        SourceLocation location, ASTNode *parent, llvm::StringRef name,
-        glu::types::FunctionTy *type, llvm::ArrayRef<ParamDecl> const params
+        llvm::BumpPtrAllocator &alloc, SourceLocation location, ASTNode *parent,
+        llvm::StringRef name, glu::types::FunctionTy *type,
+        llvm::ArrayRef<ParamDecl> const params
     )
         : DeclBase(NodeKind::FunctionDeclKind, location, parent)
         , _name(std::move(name))
         , _type(type)
-        , _body(CompoundStmt(location, {}))
+        , _body(CompoundStmt::create(alloc, location, {}))
```

Comment:
sounds wrong. the ctor should take a CompoundStmt * that is already created and filled? that's how the parser should work?